File: ktox-lua-cc/src/main/kotlin/lib/Chest.kt
package lib

import common.turtleDrop
import common.turtleGetFuelLevel
import common.turtleGetFuelLimit
import common.turtleGetItemCount
import common.turtleGetItemName
import common.turtleRefuel
import common.turtleSelect
import common.turtleSuck

// Shared home-chest servicing pattern: a fuel/supply chest at the
// turtle's start position, with overflow chests extending sideways from
// there for anything that doesn't fit. Used by any program with this
// setup (Digsite, ExcavatePro) so there's exactly one implementation of
// the chest-hopping and suck/sort logic to get right.

// +1 = turtle's original right is "sideways toward the overflow row".
// Flip to -1 if the chests turn out to be laid out the other way in-game.
const val CHEST_OVERFLOW_SIDE = 1

const val CHEST_MAX_OVERFLOW = 20

// Scratch slot used only transiently during restockChest() — never a
// permanent stockpile, since fuel is consumed immediately on pickup and
// anything else is moved to its real slot (or dropped back) the same
// visit. Slot 16 rather than something lower so it doesn't collide with
// small, deliberately low-numbered reserved slots a caller might use.
const val CHEST_INTAKE_SLOT = 16

// Deliberately restrictive, not exhaustive — turtle.refuel() itself
// already accepts anything CC:Tweaked considers valid fuel (any wood
// variant included) for free, via its own return value. This allowlist
// exists to burn LESS than that on purpose, so wood/planks/logs never
// get consumed as fuel even though they're technically valid, without
// having to enumerate every wood variant to exclude it. No blaze rod,
// per explicit request. Shared by every program with a fuel chest, so
// there's exactly one place these numbers live.
//
// Values are fuel level gained per single item burned. These match
// commonly-documented CC:Tweaked numbers but aren't verified against
// this pack/version — to confirm one yourself: hold one item, note
// turtleGetFuelLevel(), run turtle.refuel(1), and the difference is the
// exact value for your setup. Worth spot-checking coal_block and
// dried_kelp_block especially, since a wrong number there is exactly
// the near-cap waste this exists to prevent. Returns 0 for anything not
// on the allowlist (including every wood variant).
fun fuelValue(name: String): Int {
    if (name == "minecraft:charcoal") {
        return 80
    }
    if (name == "minecraft:coal") {
        return 80
    }
    if (name == "minecraft:coal_block") {
        return 800
    }
    if (name == "minecraft:lava_bucket") {
        return 1000
    }
    if (name == "minecraft:dried_kelp_block") {
        return 4000
    }
    return 0
}

// Chest N (N=0 is the home/supply chest itself, N>=1 are overflow
// chests) sits N blocks from the home chest, sideways along
// CHEST_OVERFLOW_SIDE, one row back from the turtle's home line. Must
// be called with m at the home position.
fun goToChest(m: Movement, n: Int) {
    val sideways = (m.homeHeading + CHEST_OVERFLOW_SIDE + 4) % 4
    val targetX = m.homeX + headingDx(sideways) * n
    val targetZ = m.homeZ + headingDz(sideways) * n
    navigateTo(m, targetX, m.homeY, targetZ)
    m.faceHeading((m.homeHeading + 2) % 4)
}

// True once every slot NOT exempted by `keepSlot` has something in it.
// Which slots are exempt (a fuel reserve, stock of some kind, step
// material, ...) is program-specific — see Digsite's/ExcavatePro's own
// keepSlot for what they protect and why.
fun cargoFull(keepSlot: (Int) -> Boolean): Boolean {
    var slot = 1
    var full = true
    while (slot <= 16) {
        if (!keepSlot(slot)) {
            if (turtleGetItemCount(slot) == 0) {
                full = false
            }
        }
        slot += 1
    }
    return full
}

// Dumps every non-exempt, non-empty slot into the home chest's overflow
// row, hopping to the next chest whenever the current one won't take
// any more. Must be called with m at (or able to path from) home.
fun dumpCargo(m: Movement, keepSlot: (Int) -> Boolean) {
    var chestIndex = 1
    goToChest(m, chestIndex)

    var slot = 1
    var giveUp = false
    while (slot <= 16 && !giveUp) {
        if (!keepSlot(slot)) {
            val count = turtleGetItemCount(slot)
            if (count > 0) {
                turtleSelect(slot)
                var dropped = turtleDrop(64)
                while (!dropped && !giveUp) {
                    chestIndex += 1
                    if (chestIndex > CHEST_MAX_OVERFLOW) {
                        println("All overflow chests full, stopping dump early.")
                        giveUp = true
                    } else {
                        goToChest(m, chestIndex)
                        dropped = turtleDrop(64)
                    }
                }
            }
        }
        slot += 1
    }

    navigateTo(m, m.homeX, m.homeY, m.homeZ)
    m.faceHeading(m.homeHeading)
}

// Sucks from the home chest one item at a time via CHEST_INTAKE_SLOT.
// Anything on the fuelValue() allowlist is burned as long as there's
// room under turtleGetFuelLimit() for its full value (a high-value item
// is skipped entirely rather than burned for partial value near the
// cap). Anything else is offered to `onOther(name)`, which should move
// it wherever it belongs and return true if it did.
//
// turtle.suck() always pulls whatever's in the chest's lowest-indexed
// non-empty slot — there's no way to ask for a specific item, and
// turtle.drop() puts an unwanted item right back into that same slot.
// So an item that's neither burnable nor handled by onOther gets
// dropped back, then restocking stops immediately: dropping it back
// just re-surfaces the identical stack on the very next suck(), so
// further attempts would only repeat that cycle rather than reach
// anything behind it — an unwinnable loop, not a transient hiccup.
//
// NOTE for callers writing onOther: don't write it as
// `{ name -> if (cond) { sideEffect(); true } else { false } }` — a
// multi-statement if/else branch used as an expression like that
// transpiles to invalid Lua (see AGENTS.md). Use an explicit
// `var handled = false; if (cond) { sideEffect(); handled = true };
// handled` instead.
fun restockChest(m: Movement, maxAttempts: Int, onOther: (String) -> Boolean) {
    goToChest(m, 0)
    var attempts = 0
    var chestEmpty = false
    var stuck = false
    while (attempts < maxAttempts && !chestEmpty && !stuck) {
        turtleSelect(CHEST_INTAKE_SLOT)
        val pulled = turtleSuck(64)
        if (!pulled) {
            chestEmpty = true
        } else {
            val name = turtleGetItemName(CHEST_INTAKE_SLOT)
            var handled = false
            if (name != null) {
                val value = fuelValue(name)
                if (value > 0) {
                    val headroom = turtleGetFuelLimit() - turtleGetFuelLevel()
                    if (headroom >= value) {
                        turtleSelect(CHEST_INTAKE_SLOT)
                        turtleRefuel(64)
                        handled = true
                    }
                } else if (onOther(name)) {
                    handled = true
                }
            }
            if (!handled) {
                turtleSelect(CHEST_INTAKE_SLOT)
                turtleDrop(64)
                stuck = true
                println("Chest's next item isn't usable right now (unrecognized, or would waste a high-value fuel item near the cap) - stopping restock early.")
            }
        }
        attempts += 1
    }
    navigateTo(m, m.homeX, m.homeY, m.homeZ)
    m.faceHeading(m.homeHeading)
}

File: src/pkg/player/8durt/lib/Chest.lua
-- package: lib

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "lib/Chest.kt", {["1-14"]=1,["15"]=49,["16-17"]=50,["18"]=52,["19-20"]=53,["21"]=55,["22-23"]=56,["24"]=58,["25-26"]=59,["27"]=61,["28-29"]=62,["30-35"]=64,["36"]=72,["37"]=73,["38"]=74,["39"]=75,["40-45"]=76,["46"]=84,["47"]=85,["48"]=86,["49"]=87,["50"]=88,["51-53"]=89,["54-55"]=92,["56-61"]=94,["62"]=101,["63"]=102,["64"]=104,["65"]=105,["66"]=106,["67"]=107,["68"]=108,["69"]=109,["70"]=110,["71"]=111,["72"]=112,["73"]=113,["74"]=114,["75"]=115,["76-77"]=116,["78"]=118,["79-83"]=119,["84-85"]=124,["86"]=127,["87-93"]=128,["94"]=154,["95"]=155,["96"]=156,["97"]=157,["98"]=158,["99"]=159,["100"]=160,["101"]=161,["102-103"]=162,["104"]=164,["105"]=165,["106"]=166,["107"]=167,["108"]=168,["109"]=169,["110"]=170,["111"]=171,["112"]=172,["113-115"]=173,["116-118"]=176,["119"]=179,["120"]=180,["121"]=181,["122"]=182,["123-125"]=183,["126-127"]=186,["128"]=188,["129-131"]=189}, "lib")

CHEST_OVERFLOW_SIDE = 1

CHEST_MAX_OVERFLOW = 20

CHEST_INTAKE_SLOT = 16

---@param name string
---@return number
function fuelValue(name)
    if name == "minecraft:charcoal" then
        return 80
    end
    if name == "minecraft:coal" then
        return 80
    end
    if name == "minecraft:coal_block" then
        return 800
    end
    if name == "minecraft:lava_bucket" then
        return 1000
    end
    if name == "minecraft:dried_kelp_block" then
        return 4000
    end
    return 0
end

---@param m Movement
---@param n number
function goToChest(m, n)
    local sideways = (m.homeHeading + CHEST_OVERFLOW_SIDE + 4) % 4
    local targetX = m.homeX + headingDx(sideways) * n
    local targetZ = m.homeZ + headingDz(sideways) * n
    navigateTo(m, targetX, m.homeY, targetZ)
    m:faceHeading((m.homeHeading + 2) % 4)
end

---@param keepSlot (Int) -> Boolean
---@return boolean
function cargoFull(keepSlot)
    local slot = 1
    local full = true
    while slot <= 16 do
        if not keepSlot(slot) then
            if turtle.getItemCount(slot) == 0 then
                full = false
            end
        end
        slot = ktox_plusAssign(slot, 1)
    end
    return full
end

---@param m Movement
---@param keepSlot (Int) -> Boolean
function dumpCargo(m, keepSlot)
    local chestIndex = 1
    goToChest(m, chestIndex)
    local slot = 1
    local giveUp = false
    while slot <= 16 and not giveUp do
        if not keepSlot(slot) then
            local count = turtle.getItemCount(slot)
            if count > 0 then
                turtle.select(slot)
                local dropped = turtle.drop(64)
                while not dropped and not giveUp do
                    chestIndex = ktox_plusAssign(chestIndex, 1)
                    if chestIndex > CHEST_MAX_OVERFLOW then
                        println("All overflow chests full, stopping dump early.")
                        giveUp = true
                    else
                        goToChest(m, chestIndex)
                        dropped = turtle.drop(64)
                    end
                end
            end
        end
        slot = ktox_plusAssign(slot, 1)
    end
    navigateTo(m, m.homeX, m.homeY, m.homeZ)
    m:faceHeading(m.homeHeading)
end

---@param m Movement
---@param maxAttempts number
---@param onOther (String) -> Boolean
function restockChest(m, maxAttempts, onOther)
    goToChest(m, 0)
    local attempts = 0
    local chestEmpty = false
    local stuck = false
    while attempts < maxAttempts and not chestEmpty and not stuck do
        turtle.select(CHEST_INTAKE_SLOT)
        local pulled = turtle.suck(64)
        if not pulled then
            chestEmpty = true
        else
            local name = ktoxGetItemName(CHEST_INTAKE_SLOT)
            local handled = false
            if name ~= nil then
                local value = fuelValue(name)
                if value > 0 then
                    local headroom = turtle.getFuelLimit() - turtle.getFuelLevel()
                    if headroom >= value then
                        turtle.select(CHEST_INTAKE_SLOT)
                        turtle.refuel(64)
                        handled = true
                    end
                elseif onOther(name) then
                    handled = true
                end
            end
            if not handled then
                turtle.select(CHEST_INTAKE_SLOT)
                turtle.drop(64)
                stuck = true
                println("Chest\'s next item isn\'t usable right now (unrecognized, or would waste a high-value fuel item near the cap) - stopping restock early.")
            end
        end
        attempts = ktox_plusAssign(attempts, 1)
    end
    navigateTo(m, m.homeX, m.homeY, m.homeZ)
    m:faceHeading(m.homeHeading)
end

